### ISSUE TITLE
Provide API for locking future calls to dlopen(3)

### ISSUE THREAD

**Issue by lattera** [HardenedBSD/hardenedBSD#316]
To help prevent shared object injection at runtime, provide an API that application developers could use to tell `dlopen(3)` not to allow loading shared objects from that point forward.

**Comment by lattera** [comment]
Perhaps an extension to ELF, with a PT_NOTE bit that instructs the RTLD to lock `dlopen(3)` after program initialization.

**Comment by opntr** [comment]
Why?

You can fully emulate dlopen with mmap and other code.

A such facility exists : capsicum

**Comment by lattera** [comment]
That's true, and that's one thing I plan to do with libhijack: a remote, anonymous RTLD. However, providing functionality to disable future `dlopen(3)` calls raises the bar.

Capsicum is powerful and can indeed be used to provide this type of functionality, but can be extremely difficult (and in some cases, impossible) to use--especially when dealing with existing codebases.

I mainly want to lower the bar for developers while raising the bar for attackers with respect to `dlopen(3)`.

When I implement Cross-DSO CFI support in the RTLD, I will essentially instruct `dlsym(3)` and `dlfunc(3)` to whitelist symbols that resolve to functions. This could be abused by an attacker that is able to call into the RTLD.

**Comment by sempervictus** [comment]
IIRC, the old POSIX meterpreter code tried to be a real time remote RTLD,  the weird bionic libc facilitated that somewhat, but was eventually the reason for us dropping the code altogether and Brent/Adam writing mettle on MUSL to hook the dependency layer on signals from os vs the runtime abi...

**Comment by lattera** [comment]
Yeah. I remember talking with egyp7 regarding switching the linux version of meterpreter to my libhijack tool. However, I never solved the remote RTLD problem. I'm (slowly) taking a second stab at it. Though, it relies on ptrace, and is mitigated by default on HardenedBSD (due to unprivileged process debugging disabled by default).

**Comment by sempervictus** [comment]
We would love to have libhijack integrated with the mettle codebase to permit migration and other malfeasance with "not our current pid" - I think there was even recent discussion on the subject.
